### PR TITLE
Replace _subsample with biom's optimized subsample

### PR DESCRIPTION
**Pull request by wasade**
This PR uses the `subsample` method from biom. Subsample was recently extensively optimized by @sfiligoi. And, biom as a package was made Windows compatible by @mataton. However, the new version is not yet released. This PR sets the stage for using these improvements. I'll shift to a biom release at the next opportunity.

A few important items:

- some existing tests made I think slightly conservative assumptions
- this PR is a stepping stone to providing a 2D matrix or `Table` to subsample, but in my opinion we should defer those changes as a separate PR

Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [ ] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

File: skbio/stats/_subsample.py [hunk 1]
```diff
@@ -147,19 +149,21 @@ def bin_f(x):
...
     counts : 1-D array_like
-        Vector of counts (integers) to randomly subsample from.
+        Vector of counts (integers or floats) to randomly subsample from.
     n : int
         Number of items to subsample from `counts`. Must be less than or equal
         to the sum of `counts`.
     replace : bool, optional
         If ``True``, subsample with replacement. If ``False`` (the default),
         subsample without replacement.
+    seed : int, optional
```
Comment (qiyunzhu) [inline skbio/stats/_subsample.py:165]: This `seed` may be (see [here](https://github.com/scikit-bio/scikit-bio/blob/0a9fc577ebf7ce894e3b3b67e5b3a5cb732076c5/skbio/stats/composition.py#L1880) for an example:

```
    seed : int or np.random.Generator, optional
        A user-provided random seed or random generator instance.
```

Such that a user can either provide an integer, or a pre-constructed rng. In th latter scenario, a user can do `rng = np.random.default_rng(42)` in the beginning of an analysis, and use it throughout the entire analysis.

**Comment by qiyunzhu** [comment]
@wasade Since biom-format 2.1.16 is online, would it be the time to finish and merge this PR? I was able to setup a numpy 2.0 dev environment for scikit-bio, and found that the current bottleneck is the old `_subsample` code. Thank you!

  File "skbio/stats/__subsample.pyx", line 1, in init skbio.stats.__subsample
    # ----------------------------------------------------------------------------
ImportError: numpy.core.multiarray failed to import (auto-generated because you didn't call 'numpy.import_array()' after cimporting numpy; use '<void>numpy._import_array' to disable if you are certain you don't need it).

**Comment by wasade** [comment]
@qiyunzhu, if green then I think we're good for final review

**Comment by wasade** [comment]
I don't observe the same failure on x86_64, updating CI. There may be a wider variance on the failing test than what was originally tested for.

**Comment by wasade** [comment]
Okay, we're green

**Review by qiyunzhu — COMMENTED** [review]
Looks great to me. Thanks! Just curious: What's the rationale of specifying "macos-13", "macos-14" instead of "macos-latest"?

**Comment by wasade** [comment]
Great!One is x86_64, other is M1On May 30, 2024, at 18:45, Qiyun Zhu ***@***.***> wrote:﻿
@qiyunzhu commented on this pull request.

Looks great to me. Thanks! Just curious: What's the rationale of specifying "macos-13", "macos-14" instead of "macos-latest"?

**Comment by qiyunzhu** [comment]
@wasade Interesting! I did a quick search and found this [comment](https://github.com/actions/runner-images/issues/9741#issuecomment-2075569220). If there is an official source, it will be useful to comment in the yml file. But for now, let's just merge.

**Comment by wasade** [comment]
@qiyunzhu, [here](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories)

**Pull request merged**